src/sentry/templates/sentry/toolbar/iframe.html

```diff
@@ -25,151 +27,174 @@
-            text: await response.text(),
-          };
+          dispatch[message].call(undefined, messageEvent.data);
         }
 
-        function setupMessageChannel() {
+        function getMessagePort() {
           log('setupMessageChannel()');
           const { port1, port2 } = new MessageChannel();
 
-          const messageDispatch = {
-            'log': log,
-            'request-authn': requestAuthn,
-            'clear-authn': clearAuthn,
-            'fetch': fetchProxy,
-          };
-
-          port1.addEventListener('message', messageEvent => {
+          const handlePortMessage = (messageEvent) => {
             log('port.onMessage', messageEvent.data);
 
             const { $id, message } = messageEvent.data;
-            if (!$id) {
-              return; // MessageEvent is malformed, missing $id
-            }
-
-            if (!message.$function || !(Object.hasOwn(messageDispatch, message.$function))) {
-              return; // No-op without a $function to call
+            if (!$id || !message.$function || !(Object.hasOwn(postMessageDispatch, message.$function))) {
+              return;
             }
 
-            Promise.resolve(messageDispatch[message.$function]
+            Promise.resolve(postMessageDispatch[message.$function]
               .apply(undefined, message.$args || []))
               .then($result => port1.postMessage({ $id, $result }))
               .catch(error => port1.postMessage({ $id, $error: error }));
-          });
+          };
+
+          port1.addEventListener('message', handlePortMessage);
           port1.start();
 
-          window.parent.postMessage({
-            source: 'sentry-toolbar',
-            message: 'port-connect',
-          }, referrer, [port2]);
+          return port2;
+        }
 
-          log('Sent', { message: 'port-connect', referrer });
+        function getCookieValue(cookie, domain) {
+          return `${cookie}; domain=${domain}; path=/; max-age=31536000; SameSite=none; partitioned; secure`;
         }
 
-        log('Init', { referrer, state });
+        const loginWindowMessageDispatch = {
+          'did-login': ({ cookie, token }) => {
+            if (cookie) {
+              document.cookie = getCookieValue(cookie, window.location.hostname);
+              document.cookie = getCookieValue(cookie, regionUrl);
+              log('Saved a cookie', document.cookie.indexOf(cookie) >= 0);
+            }
+            if (token) {
+              localStorage.setItem('accessToken', token);
+              log('Saved an accessToken to localStorage');
+            }
+            if (!cookie && !token) {
+              log('Unexpected: No access token found!');
+            }
+
+            postStateMessage('stale');
+          },
+        };
+
+        const parentWindowMessageDispatch = {
+          'request-login': ({ delay_ms }) => {
+            const origin = window.location.origin.endsWith('.sentry.io')
```

Comment:
Does this need to work for single-tenant or self-hosted? The `system.base-hostname` or `system.url-prefix` runtime options have the bare host, and protocol + host respectively. You could use those options to make this work in other environments.

Reply:
ideally it'll work on ST or Self-hosted.

this is here specifically because our redirect logic is messy. There's some notes here about redirects: https://github.com/getsentry/sentry/pull/80003

tl/dr: In SaaS if we include the subdomain `acme.sentry.io` in the initial request, but the user needs to login, then we'll forget the redirect target and dump them into /issues. But if we ask for sentry.io/some-page/ then the user will end up at `acme.sentry.io/some-page/` in all cases.